vendor_skins/Evercast/libobs-opengl/gl-shaderparser.c

```diff
@@ -0,0 +1,771 @@
+	dstr_cat_dstr(&glsp->gl_string, &var);
+	dstr_cat(&glsp->gl_string, "); ");
+
+	if (!cf_next_token(cfp))
+		goto fail;
+	gl_write_function_contents(glsp, &cfp->cur_token, ")");
+	dstr_cat(&glsp->gl_string, " = cos(");
+	dstr_cat_dstr(&glsp->gl_string, &var);
+	dstr_cat(&glsp->gl_string, ")");
+
+	success = true;
+
+fail:
+	dstr_free(&var);
+
+	*p_token = cfp->cur_token;
+	return success;
+}
+
+static bool gl_write_saturate(struct gl_shader_parser *glsp,
+			      struct cf_token **p_token)
+{
+	struct cf_parser *cfp = &glsp->parser.cfp;
+	cfp->cur_token = *p_token;
+
+	if (!cf_next_token(cfp))
+		return false;
+	if (!cf_token_is(cfp, "("))
+		return false;
+
+	dstr_cat(&glsp->gl_string, "clamp");
+	gl_write_function_contents(glsp, &cfp->cur_token, ")");
+	dstr_cat(&glsp->gl_string, ", 0.0, 1.0)");
+
+	*p_token = cfp->cur_token;
+	return true;
+}
+
+static inline bool gl_write_texture_call(struct gl_shader_parser *glsp,
+					 struct shader_var *var,
+					 const char *call, bool sampler)
+{
+	struct cf_parser *cfp = &glsp->parser.cfp;
+
+	if (!cf_next_token(cfp))
+		return false;
+	if (!cf_token_is(cfp, "("))
+		return false;
+
+	if (sampler) {
+		if (!cf_next_token(cfp))
+			return false;
+		const size_t sampler_id = sp_getsampler(glsp, cfp->cur_token);
+		if (sampler_id == (size_t)-1)
+			return false;
+		if (!cf_next_token(cfp))
+			return false;
+		if (!cf_token_is(cfp, ","))
+			return false;
+
+		var->gl_sampler_id = sampler_id;
+	}
+
+	dstr_cat(&glsp->gl_string, call);
+	dstr_cat(&glsp->gl_string, "(");
+	dstr_cat(&glsp->gl_string, var->name);
+	dstr_cat(&glsp->gl_string, ", ");
+	return true;
+}
+
+/* processes texture.Sample(sampler, texcoord) */
+static bool gl_write_texture_code(struct gl_shader_parser *glsp,
+				  struct cf_token **p_token,
+				  struct shader_var *var)
+{
+	struct cf_parser *cfp = &glsp->parser.cfp;
+	bool written = false;
+	cfp->cur_token = *p_token;
+
+	if (!cf_next_token(cfp))
+		return false;
+	if (!cf_token_is(cfp, "."))
+		return false;
+	if (!cf_next_token(cfp))
+		return false;
+
+	const char *function_end = ")";
+
+	if (cf_token_is(cfp, "Sample"))
+		written = gl_write_texture_call(glsp, var, "texture", true);
+	else if (cf_token_is(cfp, "SampleBias"))
+		written = gl_write_texture_call(glsp, var, "texture", true);
+	else if (cf_token_is(cfp, "SampleGrad"))
+		written = gl_write_texture_call(glsp, var, "textureGrad", true);
+	else if (cf_token_is(cfp, "SampleLevel"))
+		written = gl_write_texture_call(glsp, var, "textureLod", true);
+	else if (cf_token_is(cfp, "Load")) {
+		written = gl_write_texture_call(glsp, var, "texelFetch", false);
+		dstr_cat(&glsp->gl_string, "(");
+		function_end = ").xy, 0)";
+	}
+
+	if (!written)
+		return false;
+
+	if (!cf_next_token(cfp))
+		return false;
+
+	gl_write_function_contents(glsp, &cfp->cur_token, ")");
+	dstr_cat(&glsp->gl_string, function_end);
+
+	*p_token = cfp->cur_token;
+	return true;
+}
+
+static bool gl_write_intrinsic(struct gl_shader_parser *glsp,
```

Comment:
This function is so weird.  atan, rsqrt, mod, saturate, frac -- what is it doing I wonder... dstr_cat -- direct3d string concatenation?  Building some sort of command string to pass through somewhere for some sort of action triggering?  Strange.